tools/functional-tester/etcd-runner/main.go

```diff
@@ -212,6 +215,170 @@ func runRacer(eps []string, round int) {
 	doRounds(rcs, round)
 }
 
+func runWatcher(eps []string) {
+	// watcher for setting watch on prefixes and also handling the response on watches
+	type watcher struct {
+		prefix string
+		c      *clientv3.Client
+		wc     clientv3.WatchChan
+	}
+
+	var runningTime time.Duration = 60 // time for which operation should be performed
+	noOfPrefixes := 36                 // total number of prefixes which will be watched upon
+	watchPerPrefix := 10               // number of watchers per prefix
+	reqRate := 20                      // put request per second
+	maxKeys := 100000                  // max number of keys for put operation
+
+	ctx := context.Background()
+	prefixKeysMap := make(map[string][]string) // map for holding prefixes and values
+	ws := make([]watcher, noOfPrefixes*watchPerPrefix)
+	// pre generate prefixes equal to noOfPrefixes
+	prefixes := generateUniqueKeys(5, noOfPrefixes)
+	// pre generate keys
+	keys := generateRandomKeys(10, maxKeys)
+
+	var (
+		wg  sync.WaitGroup
+		pr  *clientv3.PutResponse
+		err error
+	)
+
+	// assign watchPerPrefix times watchers on each prefix
+	wg.Add(len(ws))
+	index := 0
+	for i := 0; i < len(ws); {
+		prefix := prefixes[index]
+		// initialize prefix key map for each prefix
+		prefixKeysMap[prefix] = make([]string, 0)
+		for j := 0; j < watchPerPrefix; j++ {
+			go func(w *watcher, prefix string) {
+				defer wg.Done()
+				w.c = randClient(eps)
+				w.prefix = prefix
+				w.wc = w.c.Watch(ctx, w.prefix, clientv3.WithPrefix())
+			}(&ws[i], prefix)
+			i++
+		}
+		index++
+	}
+	wg.Wait()
```

Comment:
> so for instance a key is put for say prefix a ax, ay, az before we start watching on it. now we start to watch on prefix a and after that keys am, an etc are put so watcher will know about am, an and keys which follow, but when we start to compare, it will start comparing with ax to am, which will fail.

etcd support watch from a particular revision. What we should do is to get the revision before we create the put routine. for every watcher, we start from that revision.